rfcs/0031-devices-api.md

```diff
@@ -0,0 +1,363 @@
+
+## Device API Functions
+In the example of Zephyr, devices are already a first class concept so many of the functions will no-op but should synchronisation be required, an example implementation could be:
+
+```c
+#include <device.h>
+
+typedef struct {
+    struct device* dev;
+    k_mutex lock;
+} tvm_device_woofles_t;
+
+int32_t TVMDeviceWooflesInit(tvm_device_woofles_t* tvm_dev, struct device* zephyr_dev) {
+    k_mutex_init(&tvm_dev->lock);
+    tvm_dev->dev = zephyr_dev;
+}
+
+int32_t TVMDeviceWooflesActivate(tvm_device_woofles_t* tvm_dev) {}
+
+int32_t TVMDeviceWooflesOpen(tvm_device_woofles_t* tvm_dev) {
+    k_mutex_lock(&tvm_dev->lock, K_FOREVER);
+}
+
+int32_t TVMDeviceWooflesClose(tvm_device_woofles_t* tvm_dev) {
+    k_mutex_unlock(&tvm_dev->lock);
+}
+
+int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev) {}
+
+int32_t TVMDeviceWooflesDestroy(tvm_device_woofles_t* tvm_dev) {
+    tvm_dev->dev = NULL;
+}
+```
+
+Whereas for CMSIS, you can use the platform-specific function to encapsulate the API to our imaginary UART accessed accelerator:
+
+```c
+typedef struct {
+    ARM_DRIVER_USART* dev;
+} tvm_device_uart_accel_t;
+
+int32_t TVMDeviceUartAccelInit(tvm_device_uart_accel_t* tvm_dev, ARM_DRIVER_USART* uart_dev) {
+    uart_dev->Initialize(NULL);
+    tvm_dev->dev = uart_dev;
+}
+
+int32_t TVMDeviceUartAccelActivate(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelOpen(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelClose(tvm_device_uart_accel_t* tvm_dev) {}
+int32_t TVMDeviceUartAccelDeactivate(tvm_device_uart_accel_t* tvm_dev) {}
+
+int32_t TVMDeviceUartAccelDestroy(tvm_device_uart_accel_t* tvm_dev) {
+    tvm_dev->dev->Uninitialize();
+}
+```
+
+## Operator Usage
+Each operator would be expected to utilise one device structure and be passed that as the `resource_handle` parameter, making the assumption that each operator or variant of an operator is only bound to one device at a time. In the following example it can be seen how a accelerators interface is implemented per platform to take this void pointer and call the platform specific driver code.
+
+```c
+// Operator takes opaque resource_handle
+int32_t my_operator(..., void* resource_handle) {
+    if (TVMDeviceWooflesInvoke(resource_handle, ...ins,outs,params...) != 0) {
+        return -1;
+    }
+}
+
+// Platform implementation
+int32_t TVMDeviceWooflesInvoke(tvm_device_woofles_t* tvm_dev) {
+    struct device* zephyr_dev = tvm_dev->dev;
+    my_accelerator_invoke(
+        zephyr_dev,
+        ...ins,outs,params...
+    );
+}
+```
+
+## PrimFunc Resource Handle
+A `tir::Var` is added to `PrimFunc` in `include/tvm/tir/function.h` which enables a `PrimFunc` to track and use the `resource_handle` parameter. This will be used by both unpacked and packed APIs to pass the resource down without packing into `TVMValue`, instead as a `void *`. 
```

Comment:
it kind of seems like we could get away without doing this by tracking which Target a function is going to run on and looking up the associated device API function in AOT/graph codegen. what do you think?

Reply:
We'd still need to pass this into the function call though to the operator though?